### PR TITLE
Drop unused arg for nvFusion

### LINKED ISSUE #482: nvfuserex has problems taking getitem.
getitem_nvfuser tests failed for a similar reason to https://github.com/Lightning-AI/lightning-thunder/blob/54bb6146ff757905925f8d9ea2197870c4971011/thunder/tests/opinfos.py#L3113-L3115. I can again create a wrapper so `slice` objects don't get passed to FusionDefinitionWrapper. But I'd love to hear thoughts from @jjsjann123 and @kevinstephano who may have a better fix.

_Originally posted by @wujingyue in https://github.com/Lightning-AI/lightning-thunder/issues/191#issuecomment-2119408081_
            

cc @tfogal

### LINKED ISSUE #531: Fusion pass should drop unused arg from parent symbols.
## 🐛 Bug

In the trace below, `ab` is a tuple of slices `(slice(3, 1), slice(1, 2))`. Currently the problem is that, during fusion pass, we make the fusion decision by looking at the nested operation `prim.slice_prim` to check if we should accept or reject the operation.

```
@torch.no_grad()
@no_autocast
def computation(t, ab):
  # t: "cuda:0 f32[5, 3]"
  # ab: "tuple"
  t0 = Tensor.__getitem__(t, ab)  # t0: "cuda:0 f32[0, 1]"
    # t0 = ltorch.getitem(t, ab)  # t0: "cuda:0 f32[0, 1]"
      # t0 = prims.slice_prim(t, [0, 1], [0, 2], [1, 1])  # t0: "cuda:0 f32[0, 1]"
  del t, ab
  return t0]
```

When we choose to fuse the `slice_prim` op, we ended up with a fusion IR like
```
@torch.no_grad()
@no_autocast
def computation(a, b):
  # t: "cuda:0 f32[5, 3]"
  # ab: "tuple"
  [t0] = nvFusion0(t, ab)
    # t0 = prims.slice_prim(t, [0, 1], [0, 2], [1, 1])  # t0: "cuda:0 f32[0, 1]"
  return t0,
```

This gives us complexity in addressing `ab` as part of the cache entry, as well as mapping `ab` to fusion inputs. We need to drop it and instead getting fusion symbol as `[t0] = nvFusion0(t)`.

### To Reproduce

```
import torch
import thunder
import operator

def foo(t, ab):
    return operator.getitem(t, ab)

# enable bookend would remove the error and let you look at the trace without fusion.
jfoo = thunder.functional.jit(foo, interpretation="python interpreter", cache="no caching", disable_torch_autograd=True, nv_enable_bookend=False)

dtype = torch.float32
t = torch.randn(5, 3, device="cuda").to(dtype=dtype)

ab = (slice(3, 1), slice(1, 2))

out = jfoo(t, ab)
out_ref = foo(t, ab)
print("\t===== prologue trace =====\n", jfoo._lc_cs.last_prologue_traces[-1])
print("\t===== compute trace =====\n", jfoo._lc_cs.last_traces)
assert(out.equal(out_ref))
```

The error below should be gone.
```
Traceback (most recent call last):
  File "/volume/thunder_517.py", line 15, in <module>
    out = jfoo(t, ab)
  File "/opt/pytorch/lightning-thunder/thunder/__init__.py", line 644, in fn_
    result = cache_entry.computation_fn(*inps)
  File "/usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py", line 115, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py", line 42, in decorate_autocast
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/amp/autocast_mode.py", line 42, in decorate_autocast
    return func(*args, **kwargs)
  File "thunder.computation_1", line 10, in computation
  File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 402, in __call__
    fd = self.get_fd(to_descriptors(args))
  File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 385, in to_descriptors
    return tuple(to_descriptor(arg) for arg in args)
  File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 385, in <genexpr>
    return tuple(to_descriptor(arg) for arg in args)
  File "/opt/pytorch/lightning-thunder/thunder/executors/nvfuserex_impl.py", line 373, in to_descriptor
    numbertype, _ = check_same_dtype(*arg)
  File "/opt/pytorch/lightning-thunder/thunder/core/utils.py", line 220, in check_same_dtype
    typ = to_dtype(a, true_dtype=True)
  File "/opt/pytorch/lightning-thunder/thunder/core/dtypes.py", line 330, in to_dtype
    raise ValueError(f"Trying to extract a dtype from object {x} with unknown type {type(x)}!")
ValueError: Trying to extract a dtype from object slice(3, 4, None) with unknown type <class 'slice'>!
```

### PR DESCRIPTION
**Pull request by kiya00**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #531 .
Fixes #482.

This PR removes the unused inputs for nvFusion during the fusion pass according to the needed_proxies dce returned

### CODE REVIEW

File: thunder/core/transform_common.py [hunk 1]
```diff
@@ -84,13 +84,18 @@ def check(inp, log_str):
 # Runs a Dead Code Elimination (DCE) pass
 # NOTE Today we are only interested in computations that produce proxies, so this will eliminate operations
 #   that only produce non-proxy objects
-def dce(trace: Trace) -> Trace:
+# NOTE needed_proxies is an in/out argument, it takes an initial set of Variables you want to keep, and return
+#   all the needed proxies of the input trace
+def dce(trace: Trace, needed_proxies=None) -> Trace:
```
Comment (jjsjann123) [inline thunder/core/transform_common.py:89]: nitpick:
```suggestion
def dce(trace: Trace, needed_proxies: None | set[Variable] =None) -> Trace:
```

File: thunder/executors/nvfuserex_impl.py [hunk 2]
```diff
@@ -619,11 +619,14 @@ def has_cuda_input_or_output(self, bsym: BoundSymbol) -> bool:
 
         return False
 
-    def _dce_bsyms(self, output, bsyms: list[BoundSymbol]) -> list[BoundSymbol]:
+    def _dce_bsyms(self, input_list, output, bsyms: list[BoundSymbol]) -> list[BoundSymbol]:
         trace = TraceCtx(None)
         trace.bound_symbols = bsyms
         bsyms.append(prims.python_return.bind(output, output=()))
-        trace = dce(trace)
+        needed_proxies: set[Variable] = set()
```
Comment (jjsjann123) [inline thunder/executors/nvfuserex_impl.py:626]: neat!

**Review by jjsjann123 — APPROVED** [review]
LGTM.

One quick comment is that, can we explicitly add a test inside `test_nvfuser.py`? Don't need anything fancy, maybe just copy over the example from #531

**Review by t-vi — APPROVED** [review]
Thank you @kiya00 @jjsjann123

**Pull request merged**